src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/SeedUserRoleAssignmentsTest.kt

```diff
@@ -0,0 +1,160 @@
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.TestInstance
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.model.SeedFileType
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.factory.StaffUserDetailsFactory
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.UserQualification
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.UserQualificationAssignmentEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.UserRole
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.UserRoleAssignmentEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.seed.CsvBuilder
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.seed.UsersSeedCsvRow
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.util.randomStringMultiCaseWithNumbers
+
+@TestInstance(TestInstance.Lifecycle.PER_METHOD)
+class SeedUserRoleAssignmentsTest : SeedTestBase() {
+  @Test
+  fun `Attempting to seed a non existent user logs an error`() {
+    mockStaffUserInfoCommunityApiCallNotFound("invalid-user")
+
+    withCsv(
+      "invalid-user",
+      userRoleAssignmentSeedCsvRowsToCsv(
+        listOf(
+          UserRoleAssignmentsSeedCsvRowFactory()
+            .withDeliusUsername("invalid-user")
+            .produce()
+        )
+      )
+    )
+
+    seedService.seedData(SeedFileType.user, "invalid-user")
+
+    assertThat(logEntries).anyMatch {
+      it.level == "error" &&
+        it.message == "Unable to complete Seed Job" &&
+        it.throwable != null &&
+        it.throwable.cause != null &&
+        it.throwable.cause!!.message == "Could not get user invalid-user"
+    }
+  }
+
+  @Test
+  fun `Attempting to seed a real but currently unknown user succeeds`() {
+    mockClientCredentialsJwtRequest()
+    mockStaffUserInfoCommunityApiCall(
+      StaffUserDetailsFactory()
+        .withUsername("unknown-user")
+        .withStaffIdentifier(6789)
+        .produce()
+    )
+
+    withCsv(
+      "unknown-user",
+      userRoleAssignmentSeedCsvRowsToCsv(
+        listOf(
+          UserRoleAssignmentsSeedCsvRowFactory()
+            .withDeliusUsername("unknown-user")
+            .withRoles(listOf(UserRole.ASSESSOR, UserRole.WORKFLOW_MANAGER))
+            .withQualifications(listOf(UserQualification.PIPE))
+            .produce()
+        )
+      )
+    )
+
+    seedService.seedData(SeedFileType.user, "unknown-user")
+
+    val persistedUser = userRepository.findByDeliusUsername("unknown-user")
+
+    assertThat(persistedUser).isNotNull
+    assertThat(persistedUser!!.deliusStaffIdentifier).isEqualTo(6789)
+    assertThat(persistedUser.roles.map(UserRoleAssignmentEntity::role)).containsExactlyInAnyOrder(
+      UserRole.ASSESSOR,
+      UserRole.WORKFLOW_MANAGER
+    )
+    assertThat(persistedUser.qualifications.map(UserQualificationAssignmentEntity::qualification)).containsExactlyInAnyOrder(
+      UserQualification.PIPE
+    )
+  }
+
```

Comment:
Do we also need a spec to describe the behaviour if the seed file contains roles or qualifications which don't exist?

I'm not sure whether:

- the process blows up on `fun addQualificationToUser(user: UserEntity, qualification: UserQualification)` in the `UserService`? 
- or does it blow up in the deserialisation test `roles = columns["roles"]!!.split(",").filter(String::isNotBlank).map(UserRole::valueOf)`?

I'm interested to understand the feedback given to the seeder. Something along the lines of "The role 'WORKSLOW_MANGER' does not exist" would be helpful...

Reply:
Would have been the latter, good point on conveying this to the user though.  Have altered to make it friendlier and added a few tests.